src/server/pps/cmds/cmds.go

```diff
@@ -619,9 +615,169 @@ each datum.`,
 
+	tailLoki := &cobra.Command{
+		Use:   "{{alias}}",
+		Short: "Tail a loki query.",
+		Long:  "Tail a loki query.",
+		Run: cmdutil.RunFixedArgs(0, func(s []string) error {
+			ctx, done := pctx.Interactive()
+			defer done()
+
+			sleep := func() {
+				select {
+				case <-ctx.Done():
+				case <-time.After(time.Second):
+				}
+			}
+
+			var connected bool
+		main:
+			for {
+				// Attempt to stay connected forever, so that tailing can continue
+				// when pachd restarts.
+				select {
+				case <-ctx.Done():
+					return nil
+				default:
+				}
+
+				client, err := pachctlCfg.NewOnUserMachine(ctx, false)
+				if err != nil {
+					if !connected {
+						// If the first connection fails, it's probably
+						// never going to work.  Give up.
+						return errors.Wrap(err, "dial")
+					}
+					log.Error(ctx, "attempt to reconnect to server failed; retrying", zap.Error(err))
+					sleep()
+					continue
+				}
+				c, err := client.TailLoki(client.Ctx(), &pps.LokiRequest{
+					Since: types.DurationProto(since),
+					Query: lokiQuery,
+				})
+				if err != nil {
+					if !connected {
+						return errors.Wrap(err, "start tail")
+					}
+					log.Error(ctx, "attempt to start tail stream failed; retrying", zap.Error(err))
+					sleep()
+					continue
+				}
+				for {
+					msg, err := c.Recv()
+					if err != nil {
+						if errors.Is(err, io.EOF) {
+							time.Sleep(time.Second)
+							continue main
+						}
+						if !connected {
+							return errors.Wrap(err, "recv")
+						}
+						log.Error(ctx, "failed to recv message; retrying", zap.Error(err))
+						sleep()
+						continue main
+					}
+					connected = true
+					switch x := msg.GetResponse().(type) {
+					case *pps.TailLokiResponse_Warning:
+						log.Info(ctx, "server warning: "+x.Warning)
+					case *pps.TailLokiResponse_Stream:
+						for _, v := range x.Stream.Values {
+							for _, m := range v.Messages {
+								if raw {
+									fmt.Println(string(m))
+									continue
+								}
+
+								t, _ := types.TimestampFromProto(v.Time)
+								var obj map[string]any
+								if err := json.Unmarshal(m, &obj); err != nil {
+									obj = map[string]any{}
+									obj["time"] = t.Format(time.RFC3339Nano)
+									obj["severity"] = "info"
+									obj["message"] = string(m)
+									if bytes.HasPrefix(m, []byte("{")) && bytes.HasSuffix(m, []byte("}")) {
+										// If the object was trying to be JSON, emit the decode error.
+										obj["decodeError"] = err.Error()
+									}
+								}
+								for k, v := range x.Stream.Metadata {
+									obj["loki."+k] = v
+								}
+								switch x.Stream.Metadata["app"] {
+								case "pachd": // We adapt other logs to pachd's format.
+								case "pachyderm-kube-event-tail", "etcd":
+									obj["time"] = obj["ts"]
+									obj["message"] = obj["msg"]
+									obj["severity"] = obj["level"]
+									delete(obj, "ts")
+									delete(obj, "msg")
+									delete(obj, "level")
+								case "pachyderm-proxy":
```

Comment:
why don't we include a "message" field for pachyderm-proxy?

Reply:
No message key in its logs.